### ISSUE TITLE
Provide official debian and RPM packages

### ISSUE THREAD

**Issue by codefromthecrypt** [envoyproxy/envoy#16867]
*Title*: Provide official debian and RPM packages

*Description*:
Right now, everything on the website mentioning "getenvoy" can be removed without breaking people.. except debian and RPM packages.

Can you please host official debian and RPM packages? Seems they can be built from the same binaries used to make docker images similar to #16830

Once this is done, hopefully there's no need for confusion anymore as getenvoy's build and packaging pipeline can quietly turn off.

*Relevant Links*:
https://github.com/envoyproxy/envoy/issues/16830

**Comment by codefromthecrypt** [comment]
cc @phlax I did a recursive grep and basically this is the last thing people can't do properly upstream.

**Comment by phlax** [comment]
im very +1 to this - and ultimately would like to create a "request to package" with debian - it really feels like envoy would be incredibly useful to end users to just be able to apt-get install

in the interim i would be very happy to create the packaging for us to host these ourselves

**Comment by phlax** [comment]
im wondering about using bazel for this - looks like a reasonable approach - although at least for debian its not clear if source packages are supoorted

in terms of prior art - tetrate's packaging is here https://github.com/tetratelabs/getenvoy-package/blob/76c310ec82aba57877acab8362ec238fe9efcb53/envoy_pkg/packages/getenvoy_package.bzl

**Comment by snowp** [comment]
I think in the past there has been some resistance towards hosting/distributing binaries (c.f. https://github.com/envoyproxy/envoy/issues/9006#issuecomment-553546959), but we could always revisit that decision. 

@envoyproxy/maintainers for visibility

**Comment by phlax** [comment]
>  I think in the past there has been some resistance towards hosting/distributing binaries (c.f. #9006 (comment)), but we could always revisit that decision.

i think that because getenvoy are no longer going to host binaries this role is shifting to CNCF

**Comment by jpeach** [comment]
FYI https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=987544

**Comment by phlax** [comment]
A quick update on this

I have a ~working prototype that builds debs/rpms for arm and x86 architectures, and tests them installed in a distro containers (#16899 )

It still needs to be figured out how this can best/most efficiently fit into the CI but it seems to mostly work so far

One issue i have is that building with rbe - there isnt an `rpmbuild` binary for rpms - but its possible that we could work around that with https://github.com/google/rpmpack

**Comment by ahmed-adam** [comment]
Hi, are there any updates on this ticket? We have also been affected by https://github.com/envoyproxy/envoy/issues/17236 except we are using RPMs.

Until this issue is resolved, does anyone have a workaround as all our builds are currently broken?  I found https://cloudsmith.io/~tetrate/repos/getenvoy-rpm-stable/packages/ - would you recommend that repo be used as a workaround?

**Comment by phlax** [comment]
cc @lizan 

not sure about the cloudsmith.io repo or the cause of https://github.com/envoyproxy/envoy/issues/17236

im working on resolving this issue but it will take some time

**Comment by phlax** [comment]
@codefromthecrypt any ideas on this ?

**Comment by codefromthecrypt** [comment]
I'll summarize: the reason for opening the issue here is "getenvoy" is going away (eg no longer maintained and we aren't supposed to use the name due). bintray is down also as the service went away.

The thing I raised last month #16868 was an interim step reduce confusion for reasons like this. Probably we should remove all references at this point. I had hoped the packages here would have landed by now and obviate the whole thing.

Folks in a pinch can use the same tarballs our tools use https://archive.tetratelabs.io/envoy/envoy-versions.json until this issue or #16830 closes.

Is there something on the packaging here that others could help complete? What's working so far? What's left to do?

**Comment by dmitri-d** [comment]
FWIW, it's possible to build rpms via fedora copr (https://copr.fedorainfracloud.org/coprs). Online builds are possible, bazel's already there (https://copr.fedorainfracloud.org/coprs/vbatts/bazel/packages/).

**Comment by codefromthecrypt** [comment]
looks like there is a draft PR out there on this topic, I had no idea.. maybe we are closer! https://github.com/envoyproxy/envoy/pull/16899

**Comment by codefromthecrypt** [comment]
should be noted that centos 7 based distributions are still widely in use, so packages should accommodate this, or some sort of clear support matrix be included to say it isn't.

**Comment by adaoud-ims** [comment]
Any advise how to build Envoy for Centos 7 ?

**Comment by gdubicki** [comment]
See https://github.com/envoyproxy/envoy-build-tools/pull/154 , @adaoud-ims .

**Comment by adaoud-ims** [comment]
> See [envoyproxy/envoy-build-tools#154](https://github.com/envoyproxy/envoy-build-tools/pull/154) , @adaoud-ims .

Thanks @gdubicki, I will have a look.

**Comment by GregTheHun** [comment]
Has there been any recent updates on this? I'd love to have this as well.

**Comment by dekimsey** [comment]
Crud. Consul 1.12 just got released and [dropped support](https://github.com/hashicorp/consul/issues/12805) for envoy 1.18. That leaves CentOS 7 without any viable envoy packages.

**Comment by JamesXNelson** [comment]
yikes... we have a bunch of centos 7 customers who could benefit from native installers... many run in sensitive data centers and don't want to install docker...  would _love_ to see this move forward. Is there anything potential contributers can do to help this along?

**Comment by mattklein123** [comment]
cc @phlax can you potentially update this issue with current status? I know this is being worked on.

**Comment by phlax** [comment]
some progress on this - we have just landed code to auto push assets on release - only envoy binaries initially -  but it will establish the pipeline

the debs are pretty much ready to publish and we can follow up fairly quickly to add - the main reason i didnt include is that they are slightly larger than expected so wanted to check out the reason before adding

there is an #envoy-release channel on slack now which i would encourage anyone interested in this to join

**Comment by codefromthecrypt** [comment]
can you describe what you mean by "auto push assets on release"? do you mean attach to GitHub release or somewhere else?

**Comment by phlax** [comment]
> do you mean attach to GitHub release or somewhere else?

github release

**Comment by codefromthecrypt** [comment]
cool. can you try to get Mac x64 and aarch64 builds up there as well as windows x64? there's been a lot of problems sourcing binaries particularly on Darwin, and we can change the sourcing accordingly.

**Comment by phlax** [comment]
could you jump on the slack channel ?

it has discussion about these and why they have not been included initially

**Comment by codefromthecrypt** [comment]
probably best to cite the permalink to here as it is the better record. I have had some difficulty getting into the envoy slack due to it restricting which domains can join. I doubt I'm the only one.

maybe you can restate if there is any plan to expand beyond that or why there won't be? regardless, this is improvement even if partially implemented, so welcome progress.

**Comment by phlax** [comment]
there is - this is just initial progress establishing the pipeline - there are various issue with all of the other binaries

>  I have had some difficulty getting into the envoy slack due to it restricting which domains can join. I doubt I'm the only one.

can you clarify the problem - im happy to follow up

**Comment by plallin** [comment]
Hello! Is there a plan to provide official FIPS-compliant packages? Currently the guidance is to build Envoy from source ([ref](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/ssl#fips-140-2)). This causes issue where there is a requirement to run CI pipelines in air-grasped environment with no Internet access as building Envoy offline is not supported either ([ref](https://github.com/envoyproxy/envoy/issues/20318)).

**Comment by phlax** [comment]
@plallin not an immediate plan - i would encourage you to join the #envoy-release slack channel as we are discussing there which binaries to add


> Envoy offline is not supported either (https://github.com/envoyproxy/envoy/issues/20318).

that bug was just closed - so that way should work now

**Comment by phlax** [comment]
...actually - that bug was partially resolved i think - you can build from a tarball now:

https://github.com/envoyproxy/envoy/blob/main/bazel/README.md#building-from-a-release-tarball

for an air-gapped build you would still need to resolve the deps somehow

**Comment by erikschul** [comment]
Bump. Tetrate no longer maintains debian or ubuntu packages ([since April 2021?](https://cloudsmith.io/~tetrate/repos/getenvoy-deb-stable/packages/)):

**Comment by yongzhang** [comment]
Bump. Latest version in Ubuntu apt source stops at 1.18.

**Comment by phlax** [comment]
the pipeline is mostly in place - i would be happy to support any effort to complete this, or review any PRs - otherwise it will have to wait until i have time available

**Comment by erikschul** [comment]
Or just recommend people to download the binary?

**Comment by phlax** [comment]
> Or just recommend people to download the binary?

the recommended, and most supported, way to run is with a container - the `distroless` image is the most lightweight and secure

the binaries in the containers are the same ones that are posted on the github release pages and the same that will be packaged in debs

running without a container means your system has to meet the glibc requirements of the static compiled bin

**Comment by codefromthecrypt** [comment]
@erikschul also unlike docker layers or the [envoy archive built on it](https://archive.tetratelabs.io/envoy/envoy-versions.json), the binaries attached to releases are not compressed. This means downloading 60M instead of 15.

That said, as this issue is titled (only debian and RPM, not windows or macOS), you could certainly use more tools using the release binary blobs instead of routing through other options to make a debian or RPM, or install manually.

**Comment by jhon-conner** [comment]
Or official binary file for centos7.

**Comment by phlax** [comment]
@jhon-conner there isnt a plan atm for adding rpms - if someone was willing to do the work i would be happy to help, but its not a priority atm

**Comment by jhon-conner** [comment]
@phlax 
😄 Thanks for your reply. Actually I compiled the source successfully after refering some documents, of course, <https://www.envoyproxy.io/docs/envoy/latest/start/building> is included. But the binary file's size is too large, 469MB.
Can you figure out how to make it smaller? Any suggestion is appreciated.
Sorry, I do not know c++. It's too hard to me.  And my command is  
```
bazel --server_javabase=/usr/lib/jvm/java-11-amazon-corretto --host_jvm_args=-Djavax.net.ssl.trustStore=$CI_PROJECT_DIR/keystore.jks --host_jvm_args=-Djavax.net.ssl.trustStorePassword=changeit test  --jobs=16 -c opt --config=sizeopt --copt=-DENVOY_IGNORE_GLIBCXX_USE_CXX11_ABI_ERROR=1 --verbose_failures --distdir=$CI_PROJECT_DIR/ext/ --distdir=$CI_PROJECT_DIR/ext2/ --distdir=$CI_PROJECT_DIR/ext4/ --repository_cache=$CI_PROJECT_DIR/ext --repository_cache=$CI_PROJECT_DIR/ext2 --repository_cache=$CI_PROJECT_DIR/ext4 --linkopt=-Wl,--strip-all //test/...
```
Yeah, as you see, I built it by gitlab ci, an environment without internet access, downloading external dependencies manually.
I put external dependencies in ext, ext2, ext4. And I used llvm.

**Comment by phlax** [comment]
@jhon-conner the easiest/most supported path for compiling is using the docker build container and the bazel targets, ie

```console
$ ./ci/run_envoy_docker.sh './ci/do_ci.sh 
```

which builds an optimized, stripped bin

the bazel target you are looking for for a single binary (non-contrib) is

```console
$ bazel build ... -c opt //source/exe:envoy-static.stripped
```

lmk if that helps - the dev docs are due some update in the near future

**Comment by jhon-conner** [comment]
@phlax 
👍 Thanks so much. I'll have a try.

**Comment by phlax** [comment]
also worth mentioning - the published Envoy bin is statically compiled so should work anywhere with a compatible glibc (not sure what reason for compiling is)

**Comment by marcindulak** [comment]
> also worth mentioning - the published Envoy bin is statically compiled so should work anywhere with a compatible glibc (not sure what reason for compiling is)

Please note that both Fedora and Debian discourage (read mostly not allow) of bundling libraries into executables. Instead the dynamic linking of the operating system libraries should be used.

The suggestion of debundling in Fedora is described here https://docs.fedoraproject.org/en-US/packaging-guidelines/#bundling. Tracking of security vulnerabilities is provided as a reason for use of system libraries instead of their copies distributed by the program (envoy) source.

 > All packages whose upstreams allow them to be built against system libraries must be built against system libraries. In this case, bundled libraries (and/or their source code) MUST be explicitly deleted during %prep

For Debian see https://blog.liw.fi/posts/2023/debian-reasons/#no-bundled-libraries

> Debian avoids using copies of libraries, or other dependencies, that are bundled with the software it packages. Many upstream projects find it easier to bundle or “vendor” dependencies, but for Debian, this means that there can be many copies of some popular libraries. When there is a need to fix a security or other severe problem in such a library, Debian would have to find all copies to fix them. This can be a lot of work, and if the security problem is urgent, it wastes valuable time to have to do that.

**Comment by phlax** [comment]
> Please note that both Fedora and Debian discourage (read mostly not allow) of bundling libraries into executables. Instead the dynamic linking of the operating system libraries should be used.

yeah - i think that is why we dont have any packages in the official distros sadly - i did reach out before to see if there was a way (either unpicking our build or having ait allowed) but iirc didnt get a response

for sure in the distro sense it makes no sense to distribute bundled bins - in other contexts - eg distroless container - it makes a lot of sense to do it this way - just as the distro want to control the deps, so do we

i would be so happy if we could find a way to bridge this gap - and would defo up for helping any efforts in this respect

**Comment by phlax** [comment]
sadly there is not a `dh_bazel` or similar afaiaa

**Comment by liudhzhyym** [comment]
Any news?

**Comment by cjonesdialpad** [comment]
Can we please get an update on this?

**Comment by phlax** [comment]
weird i thought i commented earlier - seems not

we currently publish debs to our release page

when i find some time i will add an apt repo to make those available via `apt install`

there are no current plans to publish rpms

**Comment by phlax** [comment]
as commented previously, these are statically compiled binaries - the same binaries that are available in the docker container and for direct download from the release page, just with deb packaging

**Comment by sivatarunp** [comment]
Why cant we use these packages from the github releases?
https://github.com/envoyproxy/envoy/releases/download/v1.30.1/envoy-1.30.1-linux-x86_64 ?

**Comment by phlax** [comment]
> Why cant we use these packages from the github releases?

you can - but you cant install them with dpkg/apt

**Comment by codefromthecrypt** [comment]
@phlax maybe this issue can be closed with docs

* no plans to publish deb or rpm from canonical repos
* platforms supported on release page
* how to install deb from release page
* rpm will not be released

reason is that this issue will continue to accumulate comments otherwise and I dont see any resolution besides what has been commented.

what do you think?

**Comment by sivatarunp** [comment]
There are lot of dependency issues rising when we use the releases binary. wrt gnu library
./envoy: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by ./envoy)
./envoy: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by ./envoy)
./envoy: /lib64/libpthread.so.0: version `GLIBC_2.30' not found (required by ./envoy)
./envoy: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by ./envoy)

**Comment by phlax** [comment]
@codefromthecrypt would be happy to review a docs PR

this bit is incorrect tho

> no plans to publish deb or rpm from canonical repos

there is a plan to publish a deb/apt repo, its just not that high priority right now

**Comment by phlax** [comment]
@sivatarunp you need to use a sufficiently new enough distro that has a compatible glibc

**Comment by phlax** [comment]
headsup that apt.envoyproxy.io is now live

we dont have rpm support atm, but if that was of interest to anyone i would be happy to assist/review

**Comment by matihost** [comment]
>headsup that apt.envoyproxy.io is now live

please add support to the latest Ubuntu LTS (noble)

**Comment by phlax** [comment]
the main reason for not adding more distro/versions is that it takes quite a lot of CI time to test

that said i have a vague plan to only test the earliest supported version of each distro as this would still test the deb publishing architecture and would catch incompatible glibc versions

in the meantime @matihost you can almost certainly just use the latest ubuntu that we do currently support - atm, its the same statically compiled binaries in each

**Comment by GhostLyrics** [comment]
hey @phlax!

Which strategy did you have in mind for packages in the repo? 

- one version of the binary in the repo **or** 
- all versions (including historical ones that *aren't* latest) 
 
I ask because this makes a difference for people who pin and request to install specific versions via configuration management (e.g. Ansible).

**Comment by phlax** [comment]
> Which strategy did you have in mind for packages in the repo?

im not sure im following exactly

we dont keep any binaries in the (Envoy) repo - the binaries are produced as part of the release process, the apt repo requests a list of supported versions from Envoy `main` and then constructs the apt repo from that by pulling in the released packages

> ... i have a vague plan ...

as part of our normal push/pr CI we build and test the deb packages - atm we do it for all of the supported distros - this suggestion was more about just _testing_ the earliest version of each distro rather than all (atm that would be `bullseye` and `focal`)

**Comment by GhostLyrics** [comment]
When maintaining a package repository for a Debian-like, you can generally use two approaches:

1. Only the latest version of a package stays in the repository. Debian itself does this. If other versions of a package are available, it will be because a user has configured multiple sources (e.g. a higher version of the same package may come in via the Debian Security repo)
2. All historical versions of a package stay available. Old ones are *not* purged when publishing a new one, allowing to request e.g. `envoy=1.2.3` during installation and getting exactly that one, despite it not being the latest. (e.g. GitLab does this, see below)

```
apt-cache policy gitlab-ee
gitlab-ee:
  Installed: (none)
  Candidate: 17.3.2-ee.0
  Version table:
     17.3.2-ee.0 500
        500 https://packages.gitlab.com/gitlab/gitlab-ee/debian bullseye/main arm64 Packages
     17.3.1-ee.0 500
        500 https://packages.gitlab.com/gitlab/gitlab-ee/debian bullseye/main arm64 Packages
     17.3.0-ee.0 500
        500 https://packages.gitlab.com/gitlab/gitlab-ee/debian bullseye/main arm64 Packages
     17.2.5-ee.0 500
        500 https://packages.gitlab.com/gitlab/gitlab-ee/debian bullseye/main arm64 Packages
...
```

Given these, I was wondering what you had in mind for the envoy repo.

**Comment by phlax** [comment]
ah - i see - i did ponder this, but didnt come to any firm conclusion

atm - there is no cycling of old versions - all are present - see these:

- https://apt.envoyproxy.io/pool/main/e/envoy/
- https://apt.envoyproxy.io/pool/main/e/envoy-1.28/

relatedly im wondering whether to keep no longer supported branches around - ie whether to delete 1.28 when it goes EOL (in ~6 weeks)

open to suggestions on this - even better PRs (repo is here for ref https://github.com/envoyproxy/apt)

**Comment by matihost** [comment]
>in the meantime @matihost you can almost certainly just use the latest ubuntu that we do currently support - atm, its the same statically compiled binaries in each


Yes, i can confirm that envoy package from jammy works on Ubuntu noble (24.04).

**Comment by psvmcc** [comment]
Hi, maybe use single repo like [vector](https://vector.dev/highlights/2023-11-07-new-linux-repos/) did? Without building for each distro, only for package system.

**Comment by phlax** [comment]
> ...  use single repo ...

not sure tbh

firstly i dont want to add a `script-from-interwebs-that-needs-root.sh`

then also i think the debian/ubuntu versions are significant - the currently statically compiled bin doesnt work with pre-bookworm (iirc) - so we do want to make these distro versions explicit

**Comment by GhostLyrics** [comment]
> the currently statically compiled bin doesnt work with pre-bookworm
> so we do want to make these distro versions explicit

In this case you already have exactly what's recommended.

--- 

I would, however, strongly advise against unpublishing packages. With a single exception - the content of the package being so broken that it doesn't start or destroys data - you should not unpublish software releases. In any other cases a SemVer patch-level release or similar should suffice.
Unpublishing often leads to immense breakage amongst your users. 

For example:
- if someone installs a specific version in their CI → you just broke their builds
- if someone asks for a specific version in their configuration management (e.g. Ansible, Puppet) → you just broke their playbook

I **strongly** recommend to only ever unpublish if that specific release actively harms its users. If you are pressed for disk space, consider unpublishing things that are older than 3 years, but I hesitate to even suggest that.

`<personal, subjective opinion here>` *It's really not your responsibility whether envoy users use a version that no longer officially receives any support. It's theirs - as long as you have documented what is supported at this moment and what isn't.* `</end of opinion>`

**Comment by GhostLyrics** [comment]
I see that you're trying to mix your security strategy into your availability here. I'm a bit conflicted about that, but if you insist there's also an already working example of solving that. Debian and Ubuntu do that for their outdated releases by moving them from the regular repository into a dedicated archival one ([see example](http://archive.debian.org/debian/dists/)).

Both use slightly different terminology, like this:
```diff
- deb http://deb.debian.org/debian CODENAME main
+ deb http://archive.debian.org/debian CODENAME main

- deb http://archive.ubuntu.com/ubuntu/ CODENAME main
+ deb http://old-releases.ubuntu.com/ubuntu/ CODENAME main
```

That repository (likely) has less bandwidth and is not enabled by default anywhere. Users that manually choose to use it know they are getting outdated things and unsupported things but they know the releases stay there for archival reasons.

I you'd like to make things very explicit you could use `unsupported-apt.envoyproxy.io` or `outdated-apt.envoyproxy.io`, for example. Not as pretty as `archive` but certainly makes it very clear what you're getting.

**Comment by phlax** [comment]
yeah - i considered adding a `unsupported` apt repo/s 

it would need to live under apt.envoyproxy.io tho as adding dns and/or maintaining a separate site would be non-trivial

im open to PRs to address this - the current setup wont delete anything as it stands

**Comment by GhostLyrics** [comment]
>the current setup wont delete anything as it stands

Glad to hear it!

---

If you don't want to use a subdomain for the archive repo, you *could* use the `components` field of the spec. 

```
cat debian.sources

Types: deb
# http://snapshot.debian.org/archive/debian/20240722T000000Z
URIs: http://deb.debian.org/debian
Suites: bookworm bookworm-updates
Components: main <================================================= this
Signed-By: /usr/share/keyrings/debian-archive-keyring.gpg
```

If users then want to activate it, they would change:
```diff
...
- deb[signed-by=/etc/apt/keyrings/envoy-keyring.gpg] https://apt.envoyproxy.io bookworm main
+ deb[signed-by=/etc/apt/keyrings/envoy-keyring.gpg] https://apt.envoyproxy.io bookworm main unsupported 
...
```

It would require some more configuration of your aptly repo.

**Comment by phlax** [comment]
would it not be possible to just add a directory `unsupported` and move everything under that - so users do

```
deb[signed-by=/etc/apt/keyrings/envoy-keyring.gpg] `https://apt.envoyproxy.io/unsupported bookworm main 
```

it might require some path mangling to work - not sure if it would at all

trying to think of a setup that requires minimal maintenance/complexity

i like the idea that users have to add `unsupported` somewhere to continue using the repo for old versions tho

**Comment by phlax** [comment]
by low maintenance im meaning

- doesnt require any intervention - ie automated
- doesnt add a lot of CI time - fs sizes are an issue here

tbh - the complexity issue is probs just as important, if not more

not really seeing why using a subdir is "less clean" - mho is that its more explicit - but as said above - open to PRs

on that note - we should probably be having conversations like this in the apt repo itself so future selves can see them

**Comment by phlax** [comment]
https://github.com/envoyproxy/apt/issues/23
